pubsub-mapped-api/src/test/java/com/google/pubsub/clients/consumer/KafkaConsumerTest.java

```diff
@@ -0,0 +1,352 @@
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.powermock.api.mockito.PowerMockito.when;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Empty;
+import com.google.protobuf.Timestamp;
+import com.google.pubsub.common.ChannelUtil;
+import com.google.pubsub.kafkastubs.common.KafkaException;
+import com.google.pubsub.kafkastubs.common.serialization.IntegerSerializer;
+import com.google.pubsub.kafkastubs.common.serialization.StringSerializer;
+import com.google.pubsub.kafkastubs.consumer.ConsumerRecord;
+import com.google.pubsub.kafkastubs.consumer.ConsumerRecords;
+import com.google.pubsub.v1.AcknowledgeRequest;
+import com.google.pubsub.v1.DeleteSubscriptionRequest;
+import com.google.pubsub.v1.ListTopicsRequest;
+import com.google.pubsub.v1.ListTopicsResponse;
+import com.google.pubsub.v1.PublisherGrpc.PublisherImplBase;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.PullRequest;
+import com.google.pubsub.v1.PullResponse;
+import com.google.pubsub.v1.ReceivedMessage;
+import com.google.pubsub.v1.SubscriberGrpc;
+import com.google.pubsub.v1.SubscriberGrpc.SubscriberFutureStub;
+import com.google.pubsub.v1.SubscriberGrpc.SubscriberImplBase;
+import com.google.pubsub.v1.Subscription;
+import com.google.pubsub.v1.Topic;
+import io.grpc.stub.StreamObserver;
+import io.grpc.testing.GrpcServerRule;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.regex.Pattern;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.core.classloader.annotations.SuppressStaticInitializationFor;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(Futures.class)
+@SuppressStaticInitializationFor("com.google.pubsub.common.ChannelUtil")
+public class KafkaConsumerTest {
+
+
+  @Rule
+  public final GrpcServerRule grpcServerRule = new GrpcServerRule().directExecutor();
+
+  private KafkaConsumer<Integer, String> consumer;
+
+  @Before
+  public void setUp() throws ExecutionException, InterruptedException {
+    grpcServerRule.getServiceRegistry().addService(new SubscriberImpl());
+
+    ConsumerConfig config = getConsumerConfig();
+
+    PowerMockito.mockStatic(Futures.class);
```

Comment:
Try to remove this as well; it is asking for trouble. What if gRPC internally calls Futures? It's also mocking out a really important part of your code, without verifying its behavior is correct.

As a general rule, only mock: 1) your own code or 2) interfaces.

Instead, fake/mock the service implementation; create your own SubscriberImpl that behaves how you want. You can use a different implementation in each test.

If you _really_ need to mock out the Futures call, instead create a package-private method in KafkaConsumer and override it in the test (since KafkaConsumer is extendable). If KafkaConsumer is final, then you could create a package-private interface and inject the implementation from the test.

```java
  private void unsubscribe(Collection<Subscription> subscriptions) {
    ...

    ListenableFuture<List<Empty>> listListenableFuture = futuresAllAsList(listenableFutures);
    ...
  }

  ListenableFuture<List<V>> futuresAllAsList(Iterable<? extends ListenableFuture<? extends V>> futures) {
    return Futures.allAsList(futures);
  }
```